torch/fx/symbolic_trace.py

```diff
@@ -135,11 +153,25 @@ def trace(self, root: torch.nn.Module) -> GraphModule:
         orig_call = torch.nn.Module.__call__
 
         def module_call_wrapper(mod, *args, **kwargs):
-            if not self.is_leaf_module(mod):
-                return orig_call(mod, *args, **kwargs)
-            else:
-                target = _find_module(root, mod)
-                return _create_proxy(self, 'call_module', target, args, kwargs)
+            target = _find_module(root, mod)
+
+            # Snoop up to the next `module_call_wrapper` instance in the callstack, if it exists.
+            # If that call wrapper is processing this module, and the local flag `disable_intercept`
+            # is set to True, we have already dispatched this Module invocation through `call_module`.
+            # If this is the case, `call_module` wants to invoke the module directly and not recurse
+            # through module_call_wrapper again.
+            stack = inspect.stack()
+            for frame in stack[1:]:
+                if frame.function == 'module_call_wrapper':
+                    if 'mod' in frame.frame.f_locals and frame.frame.f_locals['mod'] is mod and\
+                            frame.frame.f_locals.get('disable_intercept', False):
+                        return mod._call_impl(*args, **kwargs)
+                    else:
+                        break
```

Comment:
@zdevito  i got creative here. Let me know if there's a more sane way

Reply:
another creative idea: set `__call__` on **Module instance** for the duration of `model_call_wrapper` and then undo it back. That way you have two-level override :)

Reply:
That doesn't work, it was the first thing I tried

Reply:
Let me see if I can explain the different alternatives:
1) Patch something up in Tracer.call_module. This isn't ideal because any alternative implementations of `call_module` will need to know about this, and if they miss it they'll get weird infinite recursion errors
2) Double override like you're describing. This is fine except for when `mod` is recursive. In that case, we will lose all interception of calls except for the outermost recursive call. This can be fixed by patching something up in `Tracer.call_module`, but then we run into the same issues as (1)
3) To alleviate the common theme in (1) and (2), we could wrap `mod` in some object that will do this patching/unpatching business automatically. However, then `mod` doesn't work in `isinstance` checks
4) Directly implement this dispatch mechanism (my approach here). Basically I'm thinking about this as multi-dispatch, with an outer level for tracing, exactly like the PT dispatcher. This has been proven to work for all sorts of nested and recursive callsites, starting a new dispatch chain for each invocation. I've basically hacked this together with the `inspect` stuff, stripping away functionality if we've already done it (i.e 1 level dispatch key)

Reply:
Looking at frame information here is too complicated for what it accomplishes. Seems like it would be possible to use a 'state' inside the module_call_wrapper to track where it is being invoked from. When 'tracing' it would do its tracing behavior and invoke `call_module`, but when it invokes `call_module` it would switch the state to 'passthrough', however, when invoked in the 'passthrough' state, it would switch back to 'tracing' during the actual module call. 

I think, though, that this is getting more complicated than it deserves. We don't have a use case yet where `call_module` needs to be completely custom. Our uses cases so far such as structure-preserving tracing just need to do some special things about `is_module_leaf`. So maybe we should just not implement this yet.

Reply:
@zdevito I have heard some demand for having the `target` available in `is_leaf_module`, though, which this solves indirectly. I can put out a patch that just adds that directly